### DISCUSSION TITLE
a roadmap for dependency management via luarocks.org/rockspecs for neovim

### DISCUSSION THREAD

**Discussion by teto** [nvim-neorocks/luarocks-tag-release#73] [General]
Here is this post to outline a roadmap towards rockspec based package management for neovim plugins, and let anyone wanting to help an idea how to do so.

The goal of this organization is to promote rockspec-based installation of neovim plugins for reasons described
in these 2 posts [here](https://teto.github.io/posts/2022-06-22-neovim-plugin-luarocks-2.html) and [there](https://mrcjkb.dev/posts/2023-01-10-luarocks-tag-release.html), but mostly for automated dependency management.

Note that the neovim org is working on an alternative new) format (see https://github.com/neovim/packspec and https://packspec.org/). These are not exclusive, as long as plugins have a formal specification of their dependencies, we can convert back and forth between the different formats.

A lot of work has been done before (the luarocks-tag-release action), but here are a few things that could help with adoption (no order):

1. a script to ease the opening of PRs adding luarocks-tag-release to plugins 
2. integration of luarocks support  (remember it's just about setting `lua_modules_path = "lua"` ! probably...) in popular plugin manager (lazy.nvim for instance)
3. solve a few annoying issues related to luarocks.org:
  - https://github.com/luarocks/luarocks-site/issues/188
  - more specific tokens: at one point we wanted to test the upload of packages to luarocks in PRs (probably a bad idea but yeah) but the current tokens would allow to override the initial package in the root manifest (the global/main one). Would be nice to have a test endpoint like for cabal or removing the ability for a key to update something in the root manifest
  - not directly related but the fact that one can't update its email address on luarocks.org doesn't inspire confidence

For the last point, the ability to patch luarocks.org looks like a blocker for several people (looking at the tracker) which is why tehre is in-progress work to ease patching/testing the website ( https://github.com/teto/luarocks-site/tree/flake ).

If you wanna help with any of these points, please reach out !

Cheers

**Comment by mrcjkb** [comment]
Just dropping this note here, so I don't forget about it later:

I just remembered that there are a bunch of plugins written in languages that compile to lua (e.g. teal or fennel).
See for example https://github.com/ggandor/leap.nvim/pull/127

If the build step to compile to lua is the same or similar for many plugins, it might make sense to implement some kind of auto-detection in luarocks-tag-release (or the PR tool).

**Comment by justinmk** [comment]
> neovim org is working on an alternative new) format (see https://github.com/neovim/packspec and https://packspec.org/). These are not exclusive

100%. I see packspec (or whatever "wild west half-baked dependencies declaration format") as merely a high-leverage, low-effort way to start formally connecting the _existing web_ of vim/nvim dependencies. It doesn't preclude a luarocks approach at all, it simply formalizes what already exists--assuming it gets any traction.

> here are a few things that could help with adoption (no order):

Another major piece of friction is _installing the luarocks client_. What if we could install `luarocks` as a Nvim Lua plugin from Nvim itself? Or is there already a [bob](https://github.com/MordechaiHadad/bob) for luarocks?

**Reply by mrcjkb** [reply]
> What if we could install `luarocks` as a Nvim Lua plugin from Nvim itself?

A small pain point with `luarocks` at the moment: although it can run with `nvim` as the lua interpreter, it expects a bunch of `lua` or `luajit` library files. So currently, a lua(jit) installation is required to use `luarocks`, even if you're not using it as the interpreter.

This might be solvable with a plugin that provides these library files and configures luarocks to use them with neovim.
